sympy/utilities/lambdify.py

```diff
@@ -912,7 +912,7 @@ def _lambdifygenerated(x):
     def cleanup_linecache(filename):
         def _cleanup():
             if filename in linecache.cache:
-                del linecache.cache[filename]
+                linecache.cache[filename] = (0, None, [], filename)
```

Comment:
This brings back the memory leak from gh-27216:
```console
$ python sympy_memory.py -n 100000
Memory Usage:
0/100000: 73.30 MB
10000/100000: 75.92 MB
20000/100000: 78.67 MB
30000/100000: 81.67 MB
40000/100000: 83.67 MB
50000/100000: 86.92 MB
60000/100000: 88.92 MB
70000/100000: 91.17 MB
80000/100000: 93.17 MB
90000/100000: 97.22 MB
```

Reply:
Perhaps we simply use a try/except block instead?:
```python
try:
    del linecache.cache[filename]
except KeyError:
    pass
```
I tried that on my local branch, and it seems to work fine (as expected).
But I was never able to observe any race-conditions in the first place, @asmeurer how did you trigger the race conditions? (I tried with latest pypy, but it runs `sympy_memory.py` just fine on my linux machine).

Reply:
The race condition happened in a test failure https://github.com/sympy/sympy/pull/27223#issuecomment-2463307601

Reply:
There's no threading. It's just this function https://github.com/python/cpython/blob/bbe9b21d06c192a616bc1720ec8f7d4ccc16cab8/Lib/linecache.py#L47C5-L47C15. The code first gets a list of the `linecache` keys than iterates over those, expecting every item to be present. Somehow, specifically on PyPy, the garbage collector clears the key between those two steps. (at least that's my interpretation of the failure)